Create MigrationGuide.md for Azure.Monitor.Query (#25624)
Azure/azure-sdk-for-net: sdk/monitor/Azure.Monitor.Query/src/MigrationGuide.md

﻿# Guide for migrating from Microsoft.Azure.OperationalInsights v1.1.0 to Azure.Monitor.Query v1.0.x

This guide assists you in the migration from [Microsoft.Azure.OperationalInsights](https://www.nuget.org/packages/Microsoft.Azure.OperationalInsights/) v1.1.0 to [Azure.Monitor.Query](https://www.nuget.org/packages/Azure.Monitor.Query/) v1.0.x. Side-by-side comparisons are provided for similar operations between the two packages.

Familiarity with the `Microsoft.Azure.OperationalInsights` v1.1.0 package is assumed. If you're new to the Azure.Monitor.Query client library for .NET, see the [README for `Azure.Monitor.Query`](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.Query#readme) instead of this guide.

## Table of contents

- [Migration benefits](#migration-benefits)
    - [Cross-service SDK improvements](#cross-service-sdk-improvements)
    - [New features](#new-features)
- [Important changes](#important-changes)
    - [The client](#the-client)
    - [Client constructors and authentication](#client-constructors-and-authentication)

>     - [Send a single query request](#sending-a-single-query-request)

Reviewer: ```suggestion
    - [Send a single query request](#send-a-single-query-request)
```

To improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [.NET-specific guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html) was also introduced to ensure that .NET clients have a natural and idiomatic feel with respect to the .NET ecosystem. Further details are available in the guidelines.

### Cross-service SDK improvements

The Azure.Monitor.Query client library also takes advantage of the cross-service improvements made to the Azure development experience. Examples include:

- Using the new [`Azure.Identity`](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity#readme) library to share a single authentication approach between clients.
- A unified logging and diagnostics pipeline offering a common view of the activities across each of the client libraries.

### New features

There are a variety of new features in version 1.0 of the Azure.Monitor.Query library. Some include:

> - The ability to execute a batch of queries with the `LogsQueryClient.QueryBatch()` API.

Reviewer: ```suggestion
- The ability to execute a batch of Kusto queries with the `LogsQueryClient.QueryBatch()` API.
```

> - The ability to configure the retry policy used by the operations on the client.

Reviewer: ```suggestion
- The ability to configure the retry policy used by the operations on the client.
- The ability to map a Logs query result to a strongly typed model.
```

#### Consistency

There are now methods with similar names, signatures, and locations to create senders and receivers. The result is consistency and predictability on the various features of the library.

### Client constructors and authentication

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```.NET
using Microsoft.Azure.OperationalInsights;
using Microsoft.Rest.Azure.Authentication;

string clientId = "clientId";
string clientSecret = "clientSecret";
string domain = "domain"; //tenantId

var authEndpoint = "https://login.microsoftonline.com";
var tokenAudience = "https://api.loganalytics.io/";

var adSettings = new ActiveDirectoryServiceSettings
{
    AuthenticationEndpoint = new Uri(authEndpoint),
    TokenAudience = new Uri(tokenAudience),
    ValidateAuthority = true
};

ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(domain, clientId, clientSecret, adSettings).GetAwaiter().GetResult();

> OperationalInsightsDataClient client = new OperationalInsightsDataClient(credentials);

Reviewer: ```suggestion
var client = new OperationalInsightsDataClient(credentials);
```

﻿# Guide for migrating from Microsoft.Azure.OperationalInsights v1.1.0 to Azure.Monitor.Query v1.0.x

This guide assists you in the migration from [Microsoft.Azure.OperationalInsights](https://www.nuget.org/packages/Microsoft.Azure.OperationalInsights/) v1.1.0 to [Azure.Monitor.Query](https://www.nuget.org/packages/Azure.Monitor.Query/) v1.0.x. Side-by-side comparisons are provided for similar operations between the two client libraries.

> Familiarity with the `Microsoft.Azure.OperationalInsights` v1.1.0 package is assumed. If you're new to the `Azure.Monitor.Query` client library for .NET, see the [README for `Azure.Monitor.Query`](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.Query#readme) instead of this guide.

Reviewer: ```suggestion
Familiarity with the `Microsoft.Azure.OperationalInsights` v1.1.0 package is assumed. If you're new to the `Azure.Monitor.Query` client library for .NET, see the [README file](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.Query#readme) instead of this guide.
```

- The ability to retrieve by column name instead of by column index.
- Authentication with Azure Active Directory (Azure AD) credentials using [`Azure.Identity`](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity#readme).

For more new features, changes, and bug fixes, see the [CHANGELOG](https://github.com/Azure/azure-sdk-for-net/blob/Azure.Monitor.Query_1.0.1/sdk/monitor/Azure.Monitor.Query/CHANGELOG.md).

## Important changes

### The client

To provide a more intuitive experience, the top-level client to query logs was renamed to `LogsQueryClient` from `OperationalInsightsDataClient`. `LogsQueryClient` can be authenticated using Azure AD. This client is the single entry point to execute a single Kusto query or a batch of Kusto queries.

#### Consistency

There are now methods with similar names, signatures, and locations to create senders and receivers. The result is consistency and predictability on the various features of the library.

### Client constructors and authentication

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```csharp
using Microsoft.Azure.OperationalInsights;
using Microsoft.Rest.Azure.Authentication;

> string clientId = "clientId";
> string clientSecret = "clientSecret";
> string domain = "domain"; //tenantId
> 
> var authEndpoint = "https://login.microsoftonline.com";
> var tokenAudience = "https://api.loganalytics.io/";
> 
> var adSettings = new ActiveDirectoryServiceSettings
> {
>     AuthenticationEndpoint = new Uri(authEndpoint),
>     TokenAudience = new Uri(tokenAudience),
>     ValidateAuthority = true
> };
> 
> ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(domain, clientId, clientSecret, adSettings).GetAwaiter().GetResult();

Reviewer: Let's simplify this code sample to improve readability.
```suggestion
var adSettings = new ActiveDirectoryServiceSettings
{
    AuthenticationEndpoint = new Uri("https://login.microsoftonline.com"),
    TokenAudience = new Uri("https://api.loganalytics.io/"),
    ValidateAuthority = true
};

ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(
    "<domain>", "<clientId>", "<clientSecret>", adSettings).GetAwaiter().GetResult();
```

var client = new OperationalInsightsDataClient(credentials);
```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure.Monitor.Query;
using Azure.Identity;

var client = new LogsQueryClient(new DefaultAzureCredential());
```

### Send a single query request

> In version 1.0 of the Monitor Query library:
> 
> - The `QueryBody` is flattened. Users are expected to pass the Kusto query directly to the API.
> - The `timespan` attribute is now required, which helped to avoid querying over the entire data set.
> 

Reviewer: Lead with the client library the user is migrating from.
```suggestion
```

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```csharp

> QueryResults queryResults = client.Query($"endtoendlogs_CL | where Message == '{testIdentifierEntries}'");

Reviewer: This sample should use the same `AzureActivity` Kusto query that's used in the Monitor Query sample below.

HttpOperationResponse httpResponse = await client.QueryWithHttpMessagesAsync($"endtoendlogs_CL | where Message == '{testIdentifierEntries}'");
```

> In `Azure.Monitor.Query` v1.0.x:

Reviewer: ```suggestion
In `Azure.Monitor.Query` v1.0.x:

- The `QueryBody` is flattened. Users are expected to pass the Kusto query directly to the API.
- The `timespan` attribute is now required, which prevents querying over the entire data set.
```

```csharp

> string query = "AzureActivity | top 10 by TimeGenerated";
> Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
>     workspaceId,
>     query,
>     new QueryTimeRange(TimeSpan.FromDays(1)));

Reviewer: ```suggestion
Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
    workspaceId,
    "AzureActivity | top 10 by TimeGenerated",
    new QueryTimeRange(TimeSpan.FromDays(1)));
```

#### Consistency

There are now methods with similar names, signatures, and locations to create senders and receivers. The result is consistency and predictability on the various features of the library.

### Client constructors and authentication

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```csharp
using Microsoft.Azure.OperationalInsights;
using Microsoft.Rest.Azure.Authentication;

var adSettings = new ActiveDirectoryServiceSettings
	{
	    AuthenticationEndpoint = new Uri("https://login.microsoftonline.com"),
	    TokenAudience = new Uri("https://api.loganalytics.io/"),
	    ValidateAuthority = true
	};
	
ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(
	    "<domainId or tenantId>", "<clientId>", "<clientSecret>", adSettings).GetAwaiter().GetResult();
var client = new OperationalInsightsDataClient(credentials);
```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure.Monitor.Query;
using Azure.Identity;

var client = new LogsQueryClient(new DefaultAzureCredential());
```

### Send a single query request

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```csharp

> QueryResults queryResults = client.Query("AzureActivity | top 10 by TimeGenerated");

Reviewer: Add `using` directive to resolve `QueryResults` reference. Let's also use the async call, since the Monitor Query example is async.
```suggestion
using Microsoft.Azure.OperationalInsights.Models;

// code omitted for brevity

QueryResults response = await client.QueryAsync("AzureActivity | top 10 by TimeGenerated");
```

> HttpOperationResponse httpResponse = await client.QueryWithHttpMessagesAsync($"endtoendlogs_CL | where Message == '{testIdentifierEntries}'");

Reviewer: The `Query` example you show above is sufficient.
```suggestion
```

```

In `Azure.Monitor.Query` v1.0.x:

- The `QueryBody` is flattened. Users are expected to pass the Kusto query directly to the API.
- The `timespan` attribute is now required, which helped to avoid querying over the entire data set.

```csharp

> Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(

Reviewer: ```suggestion
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
```

	workspaceId,
	"AzureActivity | top 10 by TimeGenerated",
	new QueryTimeRange(TimeSpan.FromDays(1)));
```

### Retrieve results from a table

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```csharp

> var client = new OperationalInsightsDataClient(credentials);

Reviewer: Add `using` directives to add clarity.

```suggestion
using Microsoft.Azure.OperationalInsights;
using Microsoft.Azure.OperationalInsights.Models;

// code omitted for brevity

var client = new OperationalInsightsDataClient(credentials);
```

> var tables = client.Query("AzureActivity | top 10 by TimeGenerated").Tables;
> foreach (var table in tables)
> {
>     var rows = table.Rows;
>     foreach (var row in rows)
>     {
>         foreach (var element in row)

Reviewer: Convert to async
```suggestion
QueryResults response = await client.QueryAsync("AzureActivity | top 10 by TimeGenerated");
IList<Table> tables = response.Tables;

foreach (Table table in tables)
{
    IList<IList<string>> rows = table.Rows;
                
    foreach (IList<string> row in rows)
    {
        foreach (string element in row)
```

        {
            Console.WriteLine(element);
        }
    }
}
```

In `Azure.Monitor.Query` v1.0.x:

```csharp

> var results = await client.QueryWorkspaceAsync(

Reviewer: Add `using` directives and use the explicit type to add clarity.
```suggestion
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
```

	workspaceId,
	"AzureActivity | top 10 by TimeGenerated",
	new QueryTimeRange(TimeSpan.FromDays(1)));

> LogsTable resultTable = results.Value.Table;
> IReadOnlyList<LogsTableRow> rows = resultTable.Rows;

Reviewer: ```suggestion
LogsTable table = response.Value.Table;
IReadOnlyList<LogsTableRow> rows = table.Rows;
```

#### Consistency

There are now methods with similar names, signatures, and locations to create senders and receivers. The result is consistency and predictability on the various features of the library.

### Client constructors and authentication

In `Microsoft.Azure.OperationalInsights` v1.1.0:

```csharp
using Microsoft.Azure.OperationalInsights;
using Microsoft.Rest;
using Microsoft.Rest.Azure.Authentication;
	
// code omitted for brevity

var adSettings = new ActiveDirectoryServiceSettings
	{
	    AuthenticationEndpoint = new Uri("https://login.microsoftonline.com"),
	    TokenAudience = new Uri("https://api.loganalytics.io/"),
	    ValidateAuthority = true
	};
	
ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(
	    "<domainId or tenantId>", "<clientId>", "<clientSecret>", adSettings).GetAwaiter().GetResult();
var client = new OperationalInsightsDataClient(credentials);
```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure.Monitor.Query;
using Azure.Identity;

// code omitted for brevity

> var client = new LogsQueryClient(new DefaultAzureCredential());

Reviewer: The `Microsoft.Azure.OperationalInsights` code sample uses a tenant ID, client ID, and client secret to authenticate. Let's use `ClientSecretCredential` instead of `DefaultAzureCredential` to better show how that part translates. I think you'd have something like this:

```suggestion
var credential = new ClientSecretCredential("<domainId or tenantId>", "<clientId>", "<clientSecret>");
var client = new LogsQueryClient(credential);
```
